APPS-994 Fix bug in frag applet source code, affecting applet re-use

r-i-v-a opened the pull request.

Commit:
TODO types of fragments to cover

Commit:
TODO where fragment applet is built, where to set source code

r-i-v-a commented on compiler/src/main/scala/dx/translator/cwl/ProcessTranslator.scala:
```diff
@@ -476,6 +476,8 @@ case class ProcessTranslator(cwlBundle: CwlBundle,
       Stage(call.name, getStage(), calleeName, inputs, callee.outputVars)
     }
 
+    // TODO this is where fragment applet is built
```
For WDL, it's here
compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala

Commit:
Merge remote-tracking branch 'origin/develop' into APPS-994_frag_applet_reuse

Commit:
Where fragment applet is created for WDL

Commit:
Where block subworkflow is created

Commit:
Revert some comments

Gvaihir committed:
Merge branch 'develop' into APPS-994_frag_applet_reuse

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) CallableTranslatorTest.scala - Unit tests WIP

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) CallableTranslatorTest.scala - Unit tests WIP

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) CallableTranslatorTest.scala - Unit tests done

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) ApplicationCompiler.scala - added extracted contents for Frags/Blocks to the applet request checksum
(feat) Constants - added docContents for  ApplicationCompiler
(feat) SourceCode and case classes thereof - added extraction of source contents

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) WdlCodeGenerator.scala - Src Contents from sub blocks
(fix) CallableTranslator.scala - Frag and block document (src code or its parsed version) is now used for checksum when applets are created.
(fix) CallableTranslatorTest.scala - tests showing that docs for each block/frag don't contain the entire WF code but only a parsed version of the block/frag. Therefore, the checksum of every stage is different

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) CallableTranslatorTest.scala - tests passing

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) CompilerTest.scala - WIP

Gvaihir committed:
Merge branch 'develop' into APPS-994_frag_applet_reuse

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) RELEASE_NOTES.md - updated

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) WdlCodeGenerator.scala - minor reverts

commandlinegirl commented on compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala:
```diff
@@ -526,6 +527,8 @@ case class CallableTranslator(wdlBundle: WdlBundle,
         )
         val blockName = s"${wfName}_block_${pathStr}"
+
+        // TODO block subworkflow is created here
```
Please update the comment to sth like: A block subworkflow is created here

Gvaihir replied:
Fixed in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)

commandlinegirl commented on core/src/main/scala/dx/core/ir/package.scala:
```diff
@@ -146,6 +146,9 @@ object ExecutableKind {
   */
 trait RuntimeRequirement
 
+// TODO case class Application defined here
```
Can this be removed?

Gvaihir replied:
Fixed in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)

commandlinegirl commented on core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala:
```diff
@@ -286,6 +294,26 @@ case class CodeGenerator(typeAliases: Map[String, WdlTypes.T_Struct],
+        source = StringFileNode(contents = subBlocks.map(_.prettyFormat).mkString),
+        version = TAT.Version(outputWdlVersion)(SourceLocation.empty),
+        elements = structDefs,
+        workflow = None,
```
Should this be always None? What about a case when the block is a subworkflow?

Gvaihir replied:
AFAIK, blocks are interpreted as workflow objects on the platform anyway. However if in the `.wdl` there's an actual workflow nested in the main workflow (marked by the word `workflow` - those are parsed as workflows from the very beginning along with tasks. At the point where dependency graph is created. But I may be wrong. This code generator is acting at a much later stage, when dependency graph is constructed.

Gvaihir replied:
I retract that, fixing

Gvaihir replied:
Attempt to fix in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)
Checking

Gvaihir replied:
Fixed

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) ApplicationCompiler.scala - Checksum only includes document source contents
(fix) CallableTranslator.scala - removed comments and changed signature of createWdlBlockSrc + new tests for WF
(fix) package.scala - removed comments
(fix) WdlCodeGenerator.scala - block and WF generators get source string from source file

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) CallableTranslator.scala - reverted block to have a full WF as a document - need to work on it further

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) everywhere - minor changes

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) CallableTranslator.scala - A workflow is copied with the source attribute (see changes in TAT.Workflow) + Tests
(fix) WdlCodeGenerator.scala - refactoring
(fix) WorkflowCompiler.scala - Main change that affects a checksum calculation when deciding whether to build a new applet

Gvaihir committed:
APPS-994 Update reuse applet from task
(feat) WdlCodeGenerator.scala - doc string

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) ApplicationCompiler.scala - keeping SourceCode
(fix) WorkflowCompiler.scala - keeping SourceCode
(feat) CompilerTest.scala - WIP

Gvaihir committed:
Merge branch 'develop' into APPS-994_frag_applet_reuse

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) Compiler.scala - removing SourceCode from the applet description for schecksum. WIP
(fix) WdlCodeGenerator.scala - stand alone frag no longer needed. Refactored in CallableTranslator.scala
(fix) CallableTranslator.scala - standalone fragment is created in place
(fix) CompilerTest.scala - WIP

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) CompilerTest.scala - reverted

Gvaihir commented:
Unit tests fail because of the dependency on [wdlTools](https://github.com/dnanexus/wdlTools/pull/213) - pls approve that first

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) CompilerTest.scala - reverted

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) CompilerTest.scala - imports

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) RELEASE_NOTES.md - potential breaking change update

Gvaihir committed:
Merge branch 'develop' into APPS-994_frag_applet_reuse

# Conflicts:
#	RELEASE_NOTES.md

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) build.sbt - updated dependencies

commandlinegirl commented on RELEASE_NOTES.md:
```diff
@@ -2,7 +2,9 @@
 ## in develop
 
-* ..
+* WDL: Stages (frags/blocks) reuse applets as well. Previously only tasks were reused. **Breaking**: can break the logic
```
How about we change it to: Fragments and blocks reuse applets as well, i.e. they are not rebuilt if the code corresponding to them hasn't been updated in the WDL source file. Previously only tasks..

Gvaihir replied:
fixed in [98647f0](https://github.com/dnanexus/dxCompiler/pull/291/commits/98647f0cb1881747656c96096189f220c7a0b1d6)

commandlinegirl commented on RELEASE_NOTES.md:
```diff
@@ -2,7 +2,9 @@
-* ..
+* WDL: Stages (frags/blocks) reuse applets as well. Previously only tasks were reused. **Breaking**: can break the logic
+  of any App reuse for CWL (even tasks maybe are not reused), because `ApplicationCompiler` and `WorkflowCompiler` now
+  look at the `DocContents` for checksum, and `SourceCode` attribute is now ignored.
```
Do we need a separate CWL ticket to test this?

Gvaihir replied:
I'd fix that in a separate ticket - otherwise APPS-994 gets real big and with lots of comments

commandlinegirl replied:
That's reasonable. If you could add a note to `CWL only` https://github.com/dnanexus/dxCompiler#limitations about a limited functionality for applet reuse for CWL. Thanks!

Gvaihir replied:
added in [ab5d21b](https://github.com/dnanexus/dxCompiler/pull/291/commits/ab5d21bf743f7c619179f0c47c15839b98f53ec6)

commandlinegirl commented on compiler/src/test/resources/bugs/apps_994_v2.wdl:
```diff
@@ -0,0 +1,93 @@
+workflow reuse {
+
+	# Frag applet via sub-expression
+	# declaration of inputs is dofferent from V1. Therefore this frag should be rebuilt
```
*different

Gvaihir replied:
fixed in [98647f0](https://github.com/dnanexus/dxCompiler/pull/291/commits/98647f0cb1881747656c96096189f220c7a0b1d6)

commandlinegirl reviewed (dismissed):
LGTM modulo minor bits.

commandlinegirl commented:
@Gvaihir I've set the `cwl_conformance` label as well.

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) apps_994_v2.wdl - typos
(fix) RELEASE_NOTES.md - changed wording

Gvaihir committed:
APPS-994 Update reuse applet from task
(fix) README.md - Limitations

commandlinegirl reviewed (approved):
LGTM! Thank you!
